src/hoc/connectForm.js

```diff
@@ -29,18 +29,36 @@ const connectForm = options => Target => {
   const schemaInterface =
     schema && Object.keys(schema).length > 0 ? createSchema(schema, optionsWithDefaults.vendor) : null;
 
-  function Form({ onSubmit, onChange, onDidChange, onEvent, ...props }) {
-    const { context } = props || {};
-    const defaultValues = (schemaInterface && schemaInterface.getDefaults(context)) || {};
-
-    const [values, setValues] = useState(defaultValues);
+  function Form({
+    context = null,
+    initialValues = {},
+    resetOnInitialValuesChange = false,
+    onSubmit,
+    onChange,
+    onDidChange,
+    onEvent,
+    ...props
+  }) {
+    const defaultValues = (schemaInterface && schemaInterface.getDefaults()) || {};
```

Comment:
I think is better like we said, not spreading defaults with values, but using:
```js
// you can try and type cast objects to the defined schema
schema.cast({
  name: 'jimmy',
  age: '24',
  createdOn: '2014-09-23T19:25:25Z',
});
// => { name: 'jimmy', age: 24, createdOn: Date }
```

Reply:
something like:
```
const [values, setValues] = useState(
  (schemaInterface && schemaInterface.getInitialValues(initialValues)) || {}
);
```

Reply:
It may be an option, but the reason why I used a different approach at the very end is because:

- using cast requires a schema to be *always* defined, so the schema becomes a mandatory option;
- it may end up being potentially unsafe/dangerous since the method cast (quoting from Yup)
> Attempts to coerce the passed in value to a value that matches the schema. For example: '5' will cast to 5 when using the number() type. Failed casts generally return null, but may also return results like NaN and unexpected strings;

if said points are not an issue, I'm going to change it. Also notice that doing it this way implies that on externally forced reset the method will be called again via useEffect (useState by itself doesn't trigger a new render when the props are changed, but since it's treated like a stateful component the initial state is preserved and eventual changes are not reflected).

Reply:
ok, don't care about internal implementation, if you don't like cast.
I think you can solve point 1 with an if statement looking for schema option, and point 2 using a different method from yup, if supported.

Issue is conceptual, get defaults from schema and then call object assign with the model value is wrong, for example if you have default values based on other values, they can be overriden by our object assign.

we have to leverage the schema capabilities, if present, passing model and get back the correct values

Reply:
...with `schema.getInitialValues(myCurrentCustomer)`

the point is, validate or not validate at form init?

Reply:
I see, I was looking at it as a hierarchy:

- empty values;
- default values;
- initial values;
- internal values;

where internal values and initial values may be swapped in terms of "who wins" when it comes to what are the values to keep just by using `resetOnInitialValuesChange` property.

> for example if you have default values based on other values, they can be overridden by our object assign

for sure this may be a potential issue if the behaviour unwanted. We need to decide either if the form should be able to init anyway and discard just the wrong initial values (values not matching schema criteria) or if it should just discard them altogether whenever there's something wrong.

I would say that we need to "merge" the default values and the initial ones regardless tho. For example, imagine something similar:

```js
const schema = ({
  firstName: string().default(''),
  lastName: string(),
});

const Form = connectForm({ schema })(/* your form component */);

const FormContainer = () => {
  const initialValues = { firstName: 'Mark'};

  return <Form values={initialValues} />;
};
```

I think that in case like this, having a merge policy is somehow required.

> the point is, validate or not validate at form init?

If we decide to use a "cast" approach or any form of pre-validation before the user can actually interact within the form, then I believe we need to support it fully and allow an `validateOnInit` flag to enable/disable the option programatically. Otherwise we risk to have some wonky stuff happening like forms marked as invalid before someone even started typing anything.

Overall, I think it would a nice add-on if done right and may help to "purge" customers data and ensure more consistency in saved data.

Reply:
ok for `validateOnInit `

I think for defaults:
if you have defaults values, you have to pass a schema, no other way possible
if you don't pass a schema, values are simply `values`
Every defaults, validation is handled by schema, we have a custom schema generator with that purpose too, so you can do all the complex things you want inside your custom schema.....

Reply:
> if you have defaults values, you have to pass a schema, no other way possible

Agree on that.

> if you don't pass a schema, values are simply values

And on that too.

> Every defaults, validation is handled by schema

Here I feel a bit lost: I do agree on the fact that, if a schema is passed, the form values (whether they come from an external source - like a container - or not) should be validated against the schema sooner (`onInit`) or later (`onSubmit`). But the *defaults* values are not validated by definition, they work as initial value when nothing else is provided.

```js
const schema = object({ hairColor: string().default('blonde') });
```

this just tells that `hairColor`, if not provided, will start with `blonde` as value.

Now, my doubt is: if `hairColor` comes from an external source (redux, containers state, etc...) and it's passed as initial value to our connected form, should we just discard it if it's wrong (e.g. `hairColor = 0;`) or we discard also every other value?

I'm more into the first option (discard just the wrong value and keep everything else), but if so the values passed from the outside needs to be merged with the default values defined into the schema.

Reply:
Thinking at use-cases, Form might be rendered with remote values, unprocessed, and if validateOnInit is true, may highlights the invalid fields?
In that case initialValues are passed as is.

In "empty values" scenario, defaults might be evaluated.

So pseudo-code `if (initialValues) validate(); else setDefaults();` can be a solution?

Reply:
> Thinking at use-cases, Form might be rendered with remote values, unprocessed, and if validateOnInit is true, may highlights the invalid fields?
> In that case initialValues are passed as is.

Provided they refer to a validation schema, yes they will get highlighted.

> In "empty values" scenario, defaults might be evaluated.
> So pseudo-code `if (initialValues) validate(); else setDefaults();` can be a solution?

I'm pretty sure it should.